### DISCUSSION TITLE
deployment procedure reuse in tests

### DISCUSSION THREAD

**Discussion by wighawag** [foundry-rs/foundry#634] [General]
Hi, I love the speed of forge and its use of solidity for testing.

One thing I see missing with that though is the ability to test actual deployment setup.

In hardhat world, you can setup your deployment procedure, including the various dependencies among the smart contract as well as the arguments that depends on off-chain computation (merkle root configuration for example) that you will actually execute against a real network.

And more importantly, you do it only once and can then re-use these exact deployment procedures in your tests. (and with forking, you even run such test on the "real" network). This allow you to test the real thing and ensure the params used in your deployment make sense.

This allow your test to focus on the actual behaviour without having to instantiate each smart contract again and duplicating the deployment procedure. (doing that introduce the risk of forgetting to update them).

In hardhat world, this is achieved via js deploy script using [hardhat-deploy](https://github.com/wighawag/hardhat-deploy) test fixture that will automatically deploy the contract under test and ensure all deployment dependencies are also executed. It also benefit from snapshotting the deployment (using evm_snapshot) so the next test do not need to redeploy the whole thing and still be deterministic.

How do you imagine this in the foundry world ? 

Maybe there could be a way for deploying the contract before running the test and using snapshot to revert to that blockhain state before each test.

Like you basically execute whatever request / tx and then record the blockHash, execute test 1, revert to blockHash, execute test 2, etc....

What do you think?

**Comment by onbjerg** [comment]
This is sort of how `setUp` works - it is run before each test case, so you can do your setup in that function. Let me know if I misunderstood your question :)

**Reply by wighawag** [reply]
setUp is not what I need here. I need a way to stipulate my deployment procedure outside of solidity. Ideally I could choose my language to perform the task. Many deployment procedure require various call and read from files.

Plus in itself, setUp would not be a good way to deploy contracts themselves on a real network. It is solidity code that you would have to deploy first.

Basically in a more general sense , what I want is to be able to setup the EVM in a particular state before calling the tests. Hence why I was thinking about passing a particular blockHash for the test to revert too. But obiously we would also need a way to retrive the various contract address and extra configuration that we might want the test to have access to

**Reply by onbjerg** [reply]
If you're thinking of pulling state from a live network we do support forking mode, is that the ask?

**Reply by wighawag** [reply]
No, that's not the ask. but having forking capabilities make it even more compelling

not sure how to reformulate the ask but let me try :)

setUp is not enough when you actually want to deploy to mainnet, for various reasons including :
- setUp is solidity code, that you would have to deploy on mainnet to execute the deployment of the contracts it instantiates
- a deployment procedure might want to perform logic like reading file, constructing merkle tree, ...

this means you ll have some script written in some language that allow these features so you can actually deploy your contract on a real network

What I ask is the way to link the scripts to the tests themselves.

Because after all, without that, most of your `setUp` call is basically a replication of these deployment procedure.

What I ask is a way to remove that duplication, so you can write the deployment procedure only once and have these used in test. 

This is the killer feature of `hardhat-deploy` in my opinion (and it was the reason it was built in the first place) and it would be cool to have that same capability in foundry.

**Reply by onbjerg** [reply]
Ah! Deployment UX is being discussed in #402. Could be cool to have a feature where you could reuse that deployment in `setUp` for tests, but I also disagree that `setUp` is necessarily always a replication of the deployment procedure - if you write unit tests, `setUp` will most likely only deploy one contract instead of your entire system :smile:

**Reply by wighawag** [reply]
Cool,

Don't get me wrong, there is of course many case where `setUp` is not a replication. I never said so in my comments above. The point though is that there is cases it is and exactly at the point where you do not want to be a duplication : when you test your deployment setup.

Because in that case, you do not want to introduce false positive when you forgot to update your setUp call after changing your deployment procedure elsewhere

And with a solution like I describe, you do not even need such setUp function because the system set the test chain to have all that is needed in the first place.

**Reply by fubhy** [reply]
As mentioned in the other issue, you might be able to achieve what you are after by:

1. Spin up a local hardhat node (and run your local deployments)
2. Do a `forge test --ffi --fork-url ...` (yes, this could theoretically then be a fork-of-fork)
3. Use the `ffi` cheatcode to call a shell script and load e.g. the addresses of the contracts you deployed into your tests 

Here's how the `ffi` cheatcode works https://github.com/gakonst/foundry/blob/46e36e8aacabaf2002564d2108035557e107641b/evm-adapters/testdata/CheatCodes.sol#L179-L188: 

Here's more: https://onbjerg.github.io/foundry-book/reference/cheatcodes.html#ffi

Note that `ffi` can be a footgun ;-)
